File: algorithms/3_hard/P/00330_patching_array/greedy.c
typedef ptrdiff_t isize;

int minPatches(int* nums, int nums_len, int n_i32)
{
    const int64_t n = n_i32;
    const isize len = nums_len;
    isize idx = 0;
    int64_t miss = 1;
    int patches = 0;

    while (miss <= n) {
        if (idx < len && miss >= nums[idx]) {
            miss += nums[idx];
            idx += 1;
        } else {
            miss += miss;
            patches += 1;
        }
    }

    return patches;
}

File: algorithms/3_hard/P/00330_patching_array/greedy.rs
impl Solution {
    pub fn min_patches(nums: Vec<i32>, n: i32) -> i32 {
        let n = n as i64;
        let len = nums.len();
        let mut idx = 0_usize;
        let mut miss = 1_i64;
        let mut patches = 0_i32;

        while miss <= n {
            if idx < len && miss >= nums[idx] as i64 {
                miss += nums[idx] as i64;
                idx += 1;
            } else {
                miss += miss;
                patches += 1;
            }
        }

        patches
    }
}
